### PR TITLE
Latest posts

### PR DESCRIPTION
**Pull request by ponruslan**
develop

## GitHub Board

**Issue link**

[#33 BE - latest posts task (main page) ]( https://github.com/ita-social-projects/dokazovi-be/issues/33)

**Story link**

[#12 Story](https://github.com/ita-social-projects/dokazovi-be/issues/12)

## Code reviewers
- [x] @melnikovae87

## Summary of issue

1. - [x] DTO for posts
2. - [x] API (POST) controller with pagination (10 posts per page)
3. - [x] Service layer
4. - [x] Repository layer for latest posts and pagination
5. - [ ] Unit tests

## Summary of change

- [x]  Switch Swagger DocumentationType to OpenAPI3
- [x]  Write LatestPostDTO, PostDirectionDTO, PostTypeDTO, PostUserDTO, PostUserInstitutionDTO
- [x]  Write mappers for these DTOs.
- [x] Write endpoint to search for post by status, sorted from new to old
- [x] Add required methods in PostService and PostRepository.
- [x] Add String field "avatar" to UserEntity, add migration to it.
- [x] Delete deprecated endpoint findAll and service method for it.

### CODE REVIEW

File: src/main/java/com/softserveinc/dokazovi/dto/post/LatestPostDTO.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.softserveinc.dokazovi.dto.post;
+
+import lombok.Builder;
+import lombok.Data;
+
+import java.sql.Timestamp;
+
+@Data
+@Builder
+public class LatestPostDTO {
```
Comment (nickunderhill) [inline src/main/java/com/softserveinc/dokazovi/dto/post/LatestPostDTO.java:10]: The class name should follow naming conventions e.g. LatestPostDto

File: src/main/java/com/softserveinc/dokazovi/service/impl/PostServiceImpl.java [hunk 2]
```diff
@@ -1,23 +1,25 @@
...
-
-	public PostServiceImpl(PostRepository postRepository) {
-		this.postRepository = postRepository;
-	}
+	private final PostMapper postMapper;
 
 	@Override
-	public Page<PostEntity> findAll(Pageable pageable) {
-		return postRepository.findAll(pageable);
+	public Page<LatestPostDTO> findAllByStatus(PostStatus postStatus, Pageable pageable) {
```
Comment (nickunderhill) [inline src/main/java/com/softserveinc/dokazovi/service/impl/PostServiceImpl.java:21]: Should we return posts other than with "PUBLISHED" status for "latest" in some cases?

**Review by nickunderhill — COMMENTED** [review]
Minor changes to DTO class names.
Should discuss serving logic for "latest posts"

File: src/main/java/com/softserveinc/dokazovi/dto/post/LatestPostDTO.java (hunk 1 above)
Comment (ponruslan) [reply src/main/java/com/softserveinc/dokazovi/dto/post/LatestPostDTO.java:10]: @melnikovae87 asked us to call our DTOs

File: src/main/java/com/softserveinc/dokazovi/mapper/PostMapper.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package com.softserveinc.dokazovi.mapper;
+
+import com.softserveinc.dokazovi.dto.post.LatestPostDTO;
+import com.softserveinc.dokazovi.entity.PostEntity;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(componentModel = "spring", uses = {PostUserMapper.class, PostTypeMapper.class, PostDirectionMapper.class})
+public interface PostMapper {
```
Comment (nickunderhill) [inline src/main/java/com/softserveinc/dokazovi/mapper/PostMapper.java:9]: LatestPostMapper fits better in this context.

File: src/main/java/com/softserveinc/dokazovi/service/impl/PostServiceImpl.java (hunk 2 above)
Comment (ponruslan) [reply src/main/java/com/softserveinc/dokazovi/service/impl/PostServiceImpl.java:21]: You're right. Fixed.

Comment (nickunderhill) [reply src/main/java/com/softserveinc/dokazovi/service/impl/PostServiceImpl.java:21]: Looks good!

File: src/main/java/com/softserveinc/dokazovi/dto/post/LatestPostDTO.java (hunk 1 above)
Comment (nickunderhill) [reply src/main/java/com/softserveinc/dokazovi/dto/post/LatestPostDTO.java:10]: Ok, missed that.

File: src/main/java/com/softserveinc/dokazovi/mapper/PostMapper.java (hunk 3 above)
Comment (ponruslan) [reply src/main/java/com/softserveinc/dokazovi/mapper/PostMapper.java:9]: I think it's not a problem to map different DTOs, belonging to the same entity,  with one mapper

Comment (nickunderhill) [reply src/main/java/com/softserveinc/dokazovi/mapper/PostMapper.java:9]: Agree

**Review by nickunderhill — APPROVED** [review]

**Review by melnikovae87 — APPROVED** [review]

**Pull request merged**